### ISSUE TITLE
Focus does not return to a Project panel

### ISSUE THREAD

**Issue by zanud** [notepad-plus-plus/notepad-plus-plus#9347]
### Description of the Issue
If a Project panel has lost focus _temporarily_, then it will not get it back.

### Steps to Reproduce the Issue
There is a lot of different scenarios. For example:

**A) Operations of Notepad++**

1. Launch Notepad++.
2. Open any file.
3. Open the Project panel. (Optionally, you may create a project in it and add some files/folders to it. But this is not necessary.)
4. Invoke any operation that opens it's dialog window. The operation may be project-related or not - this is irrelevant. (For example, "Add file to the project" / "Remove file(folder) from the project" / "Remove project" / "Open Workspace" / "Save workspace as..." / "View" -> "Summary" / etc.) The only important thing - this operation must open some dialog.
5. Close the dialog by any button ("Yes" / "No" / "OK" / "Cancel" - irrelevant).

**B) Actions outside of Notepad++**

1. Launch Notepad++.
2. Open any file.
3. Open the Project panel. (Optionally, you may create a project in it and add some files/folders to it. But this is not necessary.)
4. Switch from Notepad++ to any other application (either by mouse click or by Ctrl-Tab).
5. Switch back to Notepad++ by Ctrl-Tab.

### Actual Behavior
After Step 5 focus will be on the file tab: cursor (caret) will blink in the text, and you will be able to move it by arrow keys or to type new text here.

### Expected Behavior
Focus should return to the workspace tree in the Project panel.

### Debug Information
Notepad++ v7.9.2   (32-bit)
Build time : Dec 31 2020 - 03:58:36
Path : C:\Program Files\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS Name : Microsoft Windows XP (32-bit) 
OS Build : 2600.0
Current ANSI codepage : 1251
Plugins : none

**Comment by ArkadiuszMichalski** [comment]
Temporarily, the shortcut may be used:
![image](https://user-images.githubusercontent.com/2730894/103782181-6c941700-5037-11eb-983c-930144a1ca86.png)

@sasumner This `Switch...` commands can be use as entry in context menu (they have any ids)? I don't see them directly in `View` menu.

**Comment by zanud** [comment]
> Temporarily, the shortcut may be used

Does this really works for you?
On my system after focus losing by the Project panel Ctr-Tab and Ctrl-Shift-Tab switch between opened documents (tabs) if more then 1 tab exists, and do nothing if only 1 tab is opened.

**Comment by sasumner** [comment]
>This Switch... commands can be use as entry in context menu (they have any ids)? 

I don't know; did you try/look? :-)

>I don't see them directly in View menu.

Haven't some things (intentionally without corresponding menu items) been added recently to the Shortcut Mapper, for the purpose of changing, say, the combo assigned to a toggle/select?

**Comment by zanud** [comment]
**Arkadiusz**, you forced me to discover yet one problem: if a clone an opened document in another instance of Notepad++ (by "Open in New Instance") and then open a Project panel in that new instance than this panel is almost invisible because it is only a few pixel wide.

Should I create a new issue about this?

**Comment by ArkadiuszMichalski** [comment]
>@sasumner I don't know; did you try/look? :-)

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/menuCmdID.h#L358
I tried this but not works (for example: 40000 + 4000 + 104 = 44104). I'm doing something wrong or not all IDs may be used in the context menu?

@zanud It's looks like another problem so better open separate issue for them (but I don't see such problem for me, - try describe exactly the steps to reproduce).

**Comment by zanud** [comment]
@ArkadiuszMichalski Done:
#9363

**Comment by sasumner** [comment]
The name(s) shown [HERE](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9347#issuecomment-755347718) is somewhat misleading:

Example:  I assigned a keycombo to "Switch to Function List".  
When I then press the keycbombo my already visible Function List window closes!
Pressing it again brings it back.
So it is a toggle.
But, from the text of it, I expected the keycombo to simply move focus to the Function List

**Comment by sasumner** [comment]
If I try adding one of these to the context menu, e.g. Function List, here's what appears when the menu is invoked:

![image](https://user-images.githubusercontent.com/30118311/103899991-3ef5af00-50c5-11eb-941c-799660b78860.png)

Attempting to execute it does nothing.

I did it by adding `<Item     id = "44108" />`

So it is not a good situation...but, who would ever do this?
Meaning, would even dedicated keyboardists press the context-menu key (don't know its real name) then arrow to their choice, then press Enter?
Wouldn't a dedicated keyboardist just assign a shortcut key to it and get a one-key result when wanting to switch to it?

**Comment by zanud** [comment]
@sasumner

On my system a shortcut for "Toggle Function List" opens/closes a Function List panel, and a shortcut for "Switch to Function List" only switches to the Function List panel (opening it if it is currently closed), but never closes it.

**Comment by ArkadiuszMichalski** [comment]
@sasumner
>So it is not a good situation...but, who would ever do this?
Meaning, would even dedicated keyboardists press the context-menu key (don't know its real name) then arrow to their choice, then press Enter?
Wouldn't a dedicated keyboardist just assign a shortcut key to it and get a one-key result when wanting to switch to it?

But we can set own name by `ItemNameAs`?	

> Use ItemNameAs (optional) to rename the menu item name in the context menu 
ItemNameAs can be used in any type of item. ItemNameAs value can be in any language.

Personally, I try to avoid keystrokes wherever possible, because there is too much of it for each program separately. A context menu (or button) is the perfect solution (even to not forget that some interesting option exists). 

In general, I ask because I thought that all ID could be used in the context menu or via a plugin (like Python script  to automate some actions).

**Comment by sasumner** [comment]
>But we can set own name by ItemNameAs?

So I tried putting it in the context menu by doing it by direct `<Item id = "44108" />`.
I got the result of a blank context menu item with a checkmark, (see above screenshot).
I do not think "ItemNameAs" would help, but did you try doing that? If so, what result did you get?

>Personally, I try to avoid keystrokes wherever possible, because there is too much of it for each program separately. A context menu (or button) is the perfect solution (even to not forget that some interesting option exists).

I agree with that, especially the "even to not forget..." part.

> I thought that all ID could be used in the context menu or via a plugin

I think this _should_ be the case.
Didn't @Uhf7 introduce the feature of having items without corresponding visible menu choices appear in the shortcut mapper (which feels like it could be the root of the context-menu issues we are discussing)?
Perhaps a direct question to @Uhf7 about it would help?  (Which I guess has been asked now that I have mentioned him with the `@` sign :-) )

**Comment by zanud** [comment]
> I do not think "ItemNameAs" would help, but did you try doing that?

"ItemNameAs" does allow to set a readable menu item name. But menu item itself always remains disabled.

(Funny, a menu item will be added even if "id" value is non-existent.)

**Comment by Uhf7** [comment]
I cannot reproduce part A) of the original issue on my Windows 10 1607 system. If I open a file dialog in the Project Panel, such as "Add files...", the input focus returns to the Project Panel.

I can reproduce part B) of the original issue, but I don't regard this as a problem. If Notepad++ is activated, the input focus goes to the main window, not to any side window, I'm happy with this behavior.

The "Switch to" commands are intended to
- Open the respective side window if it is not already open.
- To switch the input focus to the respective side window if it is already open.

The "Switch to" commands are intended as keyboard shortcuts only. They don't appear in the View menu, because they would blow up this menu without bringing real benefit. If you wish to switch to a side window with the mouse, simply click on it, this was the approach.

The "Switch to" commands don't work in the context menu right now, but this can be fixed by adding 3 lines in the `ContextMenu::create` function, PR follows soon. After this fix, you can include a context menu entry like this: `<Item ItemNameAs="Switch to project panel 1" id="44104"/>`, I just tested it on my system. The context menu is not prepared for commands which don't appear in the main menu.

**Comment by ArkadiuszMichalski** [comment]
@Uhf7 
Case A is true when Project panel is docked. When we open any dialog related to Project and close it then editing area becomes always active. The undocked Project window behaves properly.

>The "Switch to" commands don't work in the context menu right now, but this can be fixed by adding 3 lines in the ContextMenu::create function, PR follows soon.  

Great.

**Comment by zanud** [comment]
@ArkadiuszMichalski
> Case A is true when Project panel is docked. [...] The undocked Project window behaves properly.

Аfter your and @Uhf7 comments I checked on Windows Server 2008. On it behavior is exactly opposite: the Project panel remains active if it is docked, and loses focus if it is undocked (floated).

**Comment by Uhf7** [comment]
I can reproduce part A) of the original issue now on a Windows XP system (same OS build as posted here: https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9347#issue-777734093). 

The focus still returns to the Project Panel on my Windows 10 1607, docked and undocked.

We are currently running into another problem here. Since b58a5cc227260f7b89c3c953a1806d95f4583ada (I believe it was PR #9378), Notepad++ does not start anymore under Windows XP, because the new style file dialogs are available only since Windows Vista. Before fixing any behavior under Windows XP, we need the information, whether Notepad++ under Windows XP is now abandoned or not.

**Comment by ArkadiuszMichalski** [comment]
So we have 3 different behaviors for WinXP, Win7 and Win10. If I am not mistaken, NPP is currently only Win10 oriented.

**Comment by Uhf7** [comment]
> So we have 3 different behaviors for WinXP, Win7 and Win10.

It's even more complicated. I just tested again under Windows 7 and Windows 10.

	Windows 7:
	Notepad++ v7.9.2   (32-bit)
	Build time : Dec 31 2020 - 03:58:36
	Path : C:\Tools\Npp_792\notepad++.exe
	Admin mode : OFF
	Local Conf mode : ON
	OS Name : Windows 7 Professional (32-bit)
	OS Build : 7601.0
	Current ANSI codepage : 1252
	Plugins : mimeTools.dll NppConverter.dll NppExport.dll

	Windows 10:
	Notepad++ v7.9.2   (32-bit)
	Build time : Dec 31 2020 - 03:58:36
	Path : C:\Tools\Npp\Npp_792\notepad++.exe
	Admin mode : OFF
	Local Conf mode : ON
	OS Name : Windows 10 Pro (64-bit) 
	OS Version : 1607
	OS Build : 14393.576
	Current ANSI codepage : 1252
	Plugins : mimeTools.dll NppConverter.dll NppExport.dll 

Notepad++, Version 7.9.2, Windows 7: 
- Focus returns to the Project Panel after the "Add Files" dialog, docked and undocked.
- Focus does not return to the Project Panel after the "Add Files from Directory" dialog, if the Project Panel is docked.
- Focus returns to the Project Panel after the "Add Files from Directory" dialog, if the Project Panel is undocked.

Notepad++, current version with new folder browser, Windows 7: 
- Focus returns to the Project Panel after the "Add Files" dialog, docked and undocked.
- Focus returns to the Project Panel after the "Add Files from Directory" dialog, docked and undocked.

Notepad++, Version 7.9.2, Windows 10 1607:
- Focus returns to the Project Panel after the "Add Files" dialog, docked and undocked.
- Focus does not return to the Project Panel after the "Add Files from Directory" dialog, if the Project Panel is docked.
- Focus returns to the Project Panel after the "Add Files from Directory" dialog, if the Project Panel is undocked.

Notepad++, current version with new folder browser, Windows 10 1607: 
- Focus returns to the Project Panel after the "Add Files" dialog, docked and undocked.
- Focus returns to the Project Panel after the "Add Files from Directory" dialog, docked and undocked.

Conclusions:
- The behavior depends on the kind of dialog used. 
- The behavior under Windows 7 and Windows 10 is the same.
- The only misbehavior I see is that the focus does not return to the Project Panel after the "Add Files from Directory" dialog if the Project Panel is docked. This is fixed already in the current version.
- The only thing to do is to make the "Switch to" commands work in the context menu.
